### PR TITLE
[#1736] feat(postgresql): Support PostgreSQL index.

### PR DESCRIPTION
**Pull request by Clearvive**
### What changes were proposed in this pull request?

Support PostgreSQL index.

### Why are the changes needed?
Fix: #1736

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
UT/IT

### CODE REVIEW

File: catalogs/catalog-jdbc-postgresql/src/main/java/com/datastrato/gravitino/catalog/postgresql/operation/PostgreSqlTableOperations.java [hunk 1]
```diff
@@ -115,6 +117,39 @@ protected String generateCreateTableSql(
     return result;
   }
 
+  public static void appendIndexesSql(Index[] indexes, StringBuilder sqlBuilder) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-postgresql/src/main/java/com/datastrato/gravitino/catalog/postgresql/operation/PostgreSqlTableOperations.java:120]: could you add UT for this function?

Comment (FANNG1) [reply catalogs/catalog-jdbc-postgresql/src/main/java/com/datastrato/gravitino/catalog/postgresql/operation/PostgreSqlTableOperations.java:120]: where is the UT?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/TestPostgreSqlTableOperations.java [hunk 2]
```diff
@@ -479,4 +481,121 @@ public void testCreateAutoIncrementTable() {
...
+  public void testAppendIndexesSql() {
+    // Test append index sql success.
+    Index[] indexes =
+        new Index[] {
+          Indexes.primary("test_pk", new String[][] {{"col_1"}, {"col_2"}}),
+          Indexes.unique("u1_key", new String[][] {{"col_2"}, {"col_3"}}),
+          Indexes.unique("u2_key", new String[][] {{"col_3"}, {"col_4"}})
+        };
+    StringBuilder successBuilder = new StringBuilder();
+    PostgreSqlTableOperations.appendIndexesSql(indexes, successBuilder);
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/TestPostgreSqlTableOperations.java:576]: missing test empty index names

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/TestPostgreSqlTableOperations.java [hunk 3]
```diff
@@ -479,4 +481,121 @@ public void testCreateAutoIncrementTable() {
...
+                          "no_exist_key", new String[][] {{"no_exist_2"}, {"no_exist_3"}})
+                    }));
+    Assertions.assertTrue(
+        StringUtils.contains(
+            gravitinoRuntimeException.getMessage(),
+            "column \"no_exist_1\" named in key does not exist"));
+  }
+
+  @Test
+  public void testAppendIndexesSql() {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/TestPostgreSqlTableOperations.java:567]: seems a good start to move the test from IT to UT

File: catalogs/catalog-jdbc-postgresql/src/main/java/com/datastrato/gravitino/catalog/postgresql/operation/PostgreSqlTableOperations.java [hunk 4]
```diff
@@ -115,6 +118,40 @@ protected String generateCreateTableSql(
     return result;
   }
 
+  @VisibleForTesting
+  public static void appendIndexesSql(Index[] indexes, StringBuilder sqlBuilder) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-postgresql/src/main/java/com/datastrato/gravitino/catalog/postgresql/operation/PostgreSqlTableOperations.java:122]: remove public

File: catalogs/catalog-jdbc-postgresql/src/test/java/com/datastrato/gravitino/catalog/postgresql/operation/UTPostgreSqlTableOperations.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
+/*
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-postgresql/src/test/java/com/datastrato/gravitino/catalog/postgresql/operation/UTPostgreSqlTableOperations.java:1]: please change another name, TestXX,  UTxx is odd

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java [hunk 6]
```diff
@@ -552,4 +556,120 @@ void testCreateAndLoadSchema() {
...
+          Indexes.unique("u2_key", new String[][] {{"col_3"}, {"col_4"}}),
+          Indexes.unique("u3_key", new String[][] {{"col_5"}, {"col_4"}}),
+          Indexes.unique("u4_key", new String[][] {{"col_2"}, {"col_3"}, {"col_4"}}),
+          Indexes.unique("u5_key", new String[][] {{"col_2"}, {"col_3"}, {"col_5"}}),
+          Indexes.unique("u6_key", new String[][] {{"col_1"}, {"col_2"}, {"col_3"}, {"col_4"}}),
+        };
+
+    NameIdentifier tableIdentifier =
+        NameIdentifier.of(metalakeName, catalogName, schemaName, tableName);
+
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java:582]: please add the description of what you are testing for?

File: catalogs/catalog-jdbc-postgresql/build.gradle.kts [hunk 7]
```diff
@@ -20,6 +20,19 @@ dependencies {
   implementation(libs.commons.lang3)
   implementation(libs.commons.collections4)
   implementation(libs.jsqlparser)
+
+  testImplementation(libs.guava)
+  testImplementation(libs.commons.lang3)
+  testImplementation(libs.junit.jupiter.api)
+  testImplementation(libs.junit.jupiter.params)
+  testRuntimeOnly(libs.junit.jupiter.engine)
+  testImplementation(libs.mockito.core)
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-postgresql/build.gradle.kts:29]: do you need mock? please check the depends.

Comment (FANNG1) [reply catalogs/catalog-jdbc-postgresql/build.gradle.kts:29]: and jersey

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java [hunk 8]
```diff
@@ -552,4 +556,124 @@ void testCreateAndLoadSchema() {
...
+                  Index.IndexType.UNIQUE_KEY,
+                  null,
+                  new String[][] {{"col_1"}, {"col_3"}, {"col_4"}}),
+              Indexes.of(Index.IndexType.UNIQUE_KEY, null, new String[][] {{"col_4"}}),
+            });
+    Assertions.assertEquals(2, table.index().length);
+    Assertions.assertNotNull(table.index()[0].name());
+    Assertions.assertNotNull(table.index()[1].name());
+
+    // Test create index with same name success.
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java:662]: same column?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java [hunk 9]
```diff
@@ -552,4 +556,124 @@ void testCreateAndLoadSchema() {
...
+            new SortOrder[0],
+            new Index[] {
+              Indexes.of(
+                  Index.IndexType.UNIQUE_KEY,
+                  null,
+                  new String[][] {{"col_1"}, {"col_3"}, {"col_4"}}),
+              Indexes.of(Index.IndexType.UNIQUE_KEY, null, new String[][] {{"col_4"}}),
+            });
+    Assertions.assertEquals(2, table.index().length);
+    Assertions.assertNotNull(table.index()[0].name());
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java:659]: what is the name then?

File: catalogs/catalog-jdbc-postgresql/build.gradle.kts [hunk 10]
```diff
@@ -20,6 +20,18 @@ dependencies {
   implementation(libs.commons.lang3)
   implementation(libs.commons.collections4)
   implementation(libs.jsqlparser)
+
+  testImplementation(libs.guava)
+  testImplementation(libs.commons.lang3)
+  testImplementation(libs.junit.jupiter.api)
+  testImplementation(libs.junit.jupiter.params)
+  testRuntimeOnly(libs.junit.jupiter.engine)
+  testImplementation(libs.jersey.test.framework.core) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-postgresql/build.gradle.kts:29]: remove jersey?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 11]
```diff
@@ -239,7 +243,47 @@ protected void correctJdbcTableFields(
...
       throws SQLException {
-    return Collections.emptyList();
+    List<Index> indexes = new ArrayList<>();
+
+    // Get primary key information
+    SetMultimap<String, String> primaryKeyGroupByName = HashMultimap.create();
+    ResultSet primaryKeys = getPrimaryKeys(databaseName, tableName, metaData);
+    while (primaryKeys.next()) {
+      String columnName = primaryKeys.getString("COLUMN_NAME");
+      primaryKeyGroupByName.put(primaryKeys.getString("PK_NAME"), columnName);
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:253]: should primaryKey keep order?  sees will drop order info for current implement

Comment (FANNG1) [reply catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:253]: and we should use `KEY_SEQ` to keep primary key order 
```
KEY_SEQ short => sequence number within primary key( a value of 1 represents the first column of the primary key, a value of 2 would represent the second column within the primary key).
```

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 12]
```diff
@@ -239,7 +243,47 @@ protected void correctJdbcTableFields(
...
+    for (String key : primaryKeyGroupByName.keySet()) {
+      indexes.add(Indexes.primary(key, convertIndexFieldNames(primaryKeyGroupByName.get(key))));
+    }
+
+    // Get unique key information
+    SetMultimap<String, String> indexGroupByName = HashMultimap.create();
+    ResultSet indexInfo = getIndexInfo(databaseName, tableName, metaData);
+    while (indexInfo.next()) {
+      String indexName = indexInfo.getString("INDEX_NAME");
+      if (!indexInfo.getBoolean("NON_UNIQUE") && !primaryKeyGroupByName.containsKey(indexName)) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:264]: is it enough to only check `NON_UNIQUE`?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 13]
```diff
@@ -239,7 +243,47 @@ protected void correctJdbcTableFields(
...
+        String columnName = indexInfo.getString("COLUMN_NAME");
+        indexGroupByName.put(indexName, columnName);
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:266]: should keep order?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/bean/JdbcIndexBean.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.catalog.jdbc.bean;
+
+import com.datastrato.gravitino.rel.indexes.Index;
+import java.util.Objects;
+
+public class JdbcIndexBean {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/bean/JdbcIndexBean.java:10]: please add some description about this class

File: catalogs/catalog-jdbc-postgresql/src/test/java/com/datastrato/gravitino/catalog/postgresql/operation/UnitTestPostgreSqlTableOperations.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+package com.datastrato.gravitino.catalog.postgresql.operation;
+
+import com.datastrato.gravitino.rel.indexes.Index;
+import com.datastrato.gravitino.rel.indexes.Indexes;
+import org.apache.commons.lang3.StringUtils;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+/** Unit test for {@link PostgreSqlTableOperations}. */
+public class UnitTestPostgreSqlTableOperations {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-postgresql/src/test/java/com/datastrato/gravitino/catalog/postgresql/operation/UnitTestPostgreSqlTableOperations.java:14]: I suggest to rename this to `TestPostgreSqlTableOperations` to keep consistent with other UTs

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java [hunk 16]
```diff
@@ -552,4 +556,124 @@ void testCreateAndLoadSchema() {
...
+          Indexes.unique("u3_key", new String[][] {{"col_5"}, {"col_4"}}),
+          Indexes.unique("u4_key", new String[][] {{"col_2"}, {"col_3"}, {"col_4"}}),
+          Indexes.unique("u5_key", new String[][] {{"col_2"}, {"col_3"}, {"col_5"}}),
+          Indexes.unique("u6_key", new String[][] {{"col_1"}, {"col_2"}, {"col_3"}, {"col_4"}}),
+        };
+
+    NameIdentifier tableIdentifier =
+        NameIdentifier.of(metalakeName, catalogName, schemaName, tableName);
+
+    // Test create many indexes with same success.
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/postgresql/CatalogPostgreSqlIT.java:583]: Test create indexs with name?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 17]
```diff
@@ -239,7 +244,78 @@ protected void correctJdbcTableFields(
...
+          new JdbcIndexBean(
+              Index.IndexType.PRIMARY_KEY,
+              primaryKeys.getString("COLUMN_NAME"),
+              primaryKeys.getString("PK_NAME"),
+              primaryKeys.getInt("KEY_SEQ")));
+    }
+
+    Set<String> primaryKeyNames =
+        jdbcIndexBeans.stream()
+            .filter(jdbcIndexBean -> jdbcIndexBean.getIndexType() == Index.IndexType.PRIMARY_KEY)
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:263]: why add this filter?

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 18]
```diff
@@ -239,7 +244,78 @@ protected void correctJdbcTableFields(
...
+            .map(JdbcIndexBean::getName)
+            .collect(Collectors.toSet());
+
+    // Get unique key information
+    ResultSet indexInfo = getIndexInfo(databaseName, tableName, metaData);
+    while (indexInfo.next()) {
+      String indexName = indexInfo.getString("INDEX_NAME");
+      if (!indexInfo.getBoolean("NON_UNIQUE") && !primaryKeyNames.contains(indexName)) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:271]: why check this primaryKeyNames.contains(indexName)?

File: catalogs/catalog-jdbc-postgresql/src/test/java/com/datastrato/gravitino/catalog/postgresql/operation/UnitTestPostgreSqlTableOperations.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+  @Test
+  public void testAppendIndexesSql() {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-postgresql/src/test/java/com/datastrato/gravitino/catalog/postgresql/operation/UnitTestPostgreSqlTableOperations.java:17]: please add the test for primary key column order, same for mysql

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/CatalogMysqlIT.java [hunk 20]
```diff
@@ -754,6 +754,11 @@ public void testAutoIncrement() {
           col4,
           col5
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/catalog/jdbc/mysql/CatalogMysqlIT.java:755]: please add test column index order for mysql like pg

File: catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java [hunk 21]
```diff
@@ -239,7 +244,75 @@ protected void correctJdbcTableFields(
...
+    }
+
+    Set<String> primaryIndexNames =
+        jdbcIndexBeans.stream().map(JdbcIndexBean::getName).collect(Collectors.toSet());
+
+    // Get unique key information
+    ResultSet indexInfo = getIndexInfo(databaseName, tableName, metaData);
+    while (indexInfo.next()) {
+      String indexName = indexInfo.getString("INDEX_NAME");
+      if (!indexInfo.getBoolean("NON_UNIQUE") && !primaryIndexNames.contains(indexName)) {
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-common/src/main/java/com/datastrato/gravitino/catalog/jdbc/operation/JdbcTableOperations.java:268]: please add description why `primaryIndexNames.contains(indexName)`

**Comment by FANNG1** [comment]
LGTM except the comment

**Review by FANNG1 — APPROVED** [review]
